### ISSUE TITLE
Audio lag creeping in over time

### ISSUE THREAD

**Issue by LamptonWorm** [midwan/amiberry#821]
Hi,

I'm running Amiberry 4.1.4-rpi3-dmx on a Raspberry Pi 3b+ (Raspbian, uname reports 5.10.17).

When I have the emulation running e.g. A500 WB 1.3 (Amiga Forever ROMs) with ShufflePuck Cafe, everything starts well, but over time the audio and video become out of sync. As I play (or if I leave it running and come back later) the audio lags further and further behind the video.

I wondered if this had to do with refresh rate, I've tried various 50hz and 60hz settings in boot/config.txt but no luck. The status bar during emulation shows 50 with a red box next to it .. red isn't usually good? :)

I've tried many settings but I've been unable to keep this in sync over time, so thought I'd best post here.

Thanks,

**Comment by LamptonWorm** [comment]
Ps. using on-board 3.5m audio (will try with USB audio)

**Comment by midwan** [comment]
@LamptonWorm 
Have you tried changing the Audio Buffer size?
The default value is 8, try reducing it a bit and see if it helps. 5 seems to be a good value for RPI4, not sure how well it would behave on the RPI3.

**Comment by LamptonWorm** [comment]
Yes, I've tried a few values. Everything starts fine, but if you leave a game for say 1h and come back to it, the audio is a couple of seconds off. If you go into settings at that point and change the buffer e.g from 4 to 3 and back to 4 again, and then resume - it resets back to working. It's like it drifts over time, and touching the buffer setting resets things. Ps. is the red box next to "50" in the status bar normal?

**Comment by midwan** [comment]
@LamptonWorm 
The color is identical to WinUAE, so that's fine, yes. :)
I'd have to test this to see if I can recreate it. Does it only happen with this specific game? What about others?
Did you try an ADF version? Or WHDLoad?

**Comment by LamptonWorm** [comment]
I've only tried (and noticed the same) with ShufflePuck and OctaMed V4. I use the latter the most, and that is where I noticed things initially. Both are ADF. Thanks for confirming about the lights. It is strange, and I am tempted to start a completely new install, OS + Amiberry, and see if it still happens. Ps. I had the same issue with both onboard audio and USB. Is it using Alsa btw?

**Comment by midwan** [comment]
It's not hardcoded to use any specific subsystem. It will use whatever is available on your system, through SDL2's mechanism.

**Comment by LamptonWorm** [comment]
I did a fresh OS install to a new card and downloaded the RPI3 DMX binaries, I get the same behavior. It isn't immediately obvious -- it has to run for a while, but not too long, before it is noticeable. I tried with different audio settings in Amiberry but no luck. If someone else can reproduce it, it helps validate it is not something this end. With the same roms/adf in WinUAE all is well.

**Comment by midwan** [comment]
I did a first test today, but i could not recreate the problem.
I tested it with Retropie, using the latest Amiberry version from master (v4.1.4 manually compiled), the Whdload version of Shufflepuck Cafe.
I played a bit, then let it sit for more than 2 hours, then i played some more.

No audio lag was noticeable for me.

The config I used was slightly modified from what the XML auto detected, due to gfx glitches with those settings. I changed CPU speed to 7Mhz to fix that. I also changed the audio buffer to 5. All changes were made only once, at the beginning of booting the game.

**Comment by LamptonWorm** [comment]
Thanks for trying. I think unless someone else encounters/reports this as an issue (or I can work out this end how to resolve/introduce it on demand) there isn't much more to do but thanks for the efforts. Will post if I stumble upon anything.

**Comment by Drag0nFly** [comment]
I've also noticed audio delays, however they appear quite consistent at approx ~1sec. Which setting is the “audio buffer” that is referred to above, btw.? (mentioned as being set to "8" (kb?) by default)

In both the default.uae, rp-a500.uae, rp-a1200.uae and the main amiberry.conf itself the default on my Retroarch/pie install the default is 16384 (16kB)
`grep -i buffer /opt/retropie/configs/amiga/amiberry/conf/amiberry.conf`
`default_sound_buffer=16384`

`grep sound_max_buff /opt/retropie/configs/amiga/amiberry/conf/*uae`
`/opt/retropie/configs/amiga/amiberry/conf/default.uae:sound_max_buff=16384`
`/opt/retropie/configs/amiga/amiberry/conf/rp-a500.uae:sound_max_buff=16384?`
`/opt/retropie/configs/amiga/amiberry/conf/rp-a1200.uae:sound_max_buff=16384`

Reducing to 4kB did have **some** effect, but there is still a delay present. Wonder if this has to do with the Pulseaudio backend which Retropie uses (it least it caused all sorts of havoc on a HTC system before it was removed for good)

**Comment by DJP75** [comment]
I have the exact same problem: audio lags over time! In using amiberry on dietpi.

**Comment by PeterBarrack** [comment]
I also see audio drift over time, but only when audio buffer is set to "Push" - it takes 5 minutes to be noticable and 5 more minutes to be very obvious.

I'm running a 50hz display on a Pi 400 with SDL2.

I've tested the issue on 4.1.4 and 4.1.5RC2 on many configurations - fastest Amiga to slowest, ADFs, HDFs, WHDLoad, etc, and the behaviour is consistent.

I find "Push" with a buffer of 1 gives perfect-sounding (but drifting over time) audio whereas "Pull" with a buffer of 4, 5 or 6 gives the odd audio glitch (but with no drift), especially in demos such as a Cebit90. I tend to use "Pull" because over time "Push" will drift out of time, which on games such as Sensible Soccer is very obvious.

**Comment by midwan** [comment]
Does it work better if you use Push with a buffer of 4?

**Comment by PeterBarrack** [comment]
Hi @midwan, 

I tested it using Push with a buffer of 4 and it's still the same. This video shows the issue after about 10-15 minutes of running Sensible Soccer using Push/4 : https://photos.app.goo.gl/pJdxsWi8xxQsUNZ99

It can be hard to tell from just the ball-hit noise, but at around 25 seconds a goal is nearly scored and crowd noise comes a couple of seconds late. Also near the end of the clip a goal is scored and the cheering starts a couple of seconds after rather than immediately.

**Comment by midwan** [comment]
Yes, if you change the audio buffer size then the buffer is reset and it re-syncs...

@ogotai Does switching mode to Pull make any difference for you?

**Comment by solskogen** [comment]
@ogotai Do you use a controller or are you emulating a joystick with the keyboard?

**Comment by ogotai** [comment]
@midwan so far so good - changed mode to Pull and set buffer size to 6 and this annoying lag can be heard no more. I'll keep testing. Funny that before I posted my first message in this thread I also lowered emulation accuracy by one step and it took much longer time for sound lag to develop :) Anyway it's all good for now so thanks!

@solskogen I do actually use a X360 compatible gamepad (Genesis P65) which I can switch from Xinput to Dinput if required. Both settings had no effect on sound lag issue. Also unplugging it and using keyboard instead didn't help at all, only changing sound mode to Pull.

**Comment by solskogen** [comment]
Good to know. I wonder if the audio delay happens even if you're not playing. I managed to reproduce it once the same problem when I was playing Pang yesterday, but not when I tried the second time not playing (Pang has a demo when you leave it)

**Comment by PeterBarrack** [comment]
I've not found any difference between playing or not playing. I have two old Zipstiks plugged into the Pi 400 via a DB9-GPIO adapter, using the db9_gpio_rpi driver. The video in my previous comment shows a CPU vs CPU Sensible Soccer match being played, during which the sound lag gradually worsens. I've also seen the same behaviour in other games.

**Comment by midwan** [comment]
@PeterBarrack 
But I assume in your case also, that if you use Pull with a buffer or around 6, it works without that problem?
Just trying to piece together what we know so far...

**Comment by PeterBarrack** [comment]
Hi @midwan ,
Yes, I usually use Pull with a buffer of 4, and that removes the lag issue. I have also used Pull with buffer of 5 and 6 in the past which also works fine - no lag.

**Comment by midwan** [comment]
I'm adding an option in `amiberry.conf` to handle the default behavior of the emulator, if there's no config value to override this of course: https://github.com/midwan/amiberry/wiki/Amiberry.conf-options

I'm also changing the default option from Push to Pull, and increasing the default value for the sound buffer slightly.

All these options can be handled from the `amiberry.conf` file, which is also being updated to include them with the defaults.
Of course, if you already have configurations saved they are not affected. This is only what the emulator will use when it starts up.

Hopefully this will fix the immediate problem for those facing it, and we can see about improvements in this area in the future.
